yorkie/src/main/kotlin/dev/yorkie/util/SplayTreeSet.kt

```diff
@@ -0,0 +1,364 @@
+package dev.yorkie.util
+
+import com.google.common.annotations.VisibleForTesting
+
+/**
+ * SplayTreeSet is weighted binary search tree set based on Splay tree.
+ * original paper on Splay Trees:
+ * @link https://www.cs.cmu.edu/~sleator/papers/self-adjusting.pdf
+ */
+// should SplayTree implement MutableSet?
+internal class SplayTreeSet<V>(private val lengthCalculator: (V) -> Int = { 1 }) {
```

Comment:
I think that the idea, of dealing with creation-time-based sets is novel. 😃
Are there any constraints on becoming a TreeSet's Value?

Reply:
Only constraint that comes to mind is that it doesn't allow duplicates.
Therefore if the caller wants use it with duplicated primitive values they need to create a wrapper which is a little memory-inefficient.
But since we are planning to use this for objects with a unique TimeTicket, I think we have no problem :)